// playwright-cli/lib/traceViewer/web/components/actionRegistry.js
"use strict";
/*
  Copyright (c) Microsoft Corporation.
 
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at
 
      http://www.apache.org/licenses/LICENSE-2.0
 
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.actionRegistry = exports.Action = void 0;
const events_1 = require("./events");
const multimap_1 = require("./multimap");
class Action {
    constructor(options) {
        this._enabled = true;
        this._toggled = false;
        this._onUpdatedEmitter = new events_1.EventEmitter();
        this.onUpdated = this._onUpdatedEmitter.event;
        this._options = { ...options };
    }
    get label() {
        return this._options.label || '';
    }
    get title() {
        return this._options.title || '';
    }
    get icon() {
        return this._options.icon ? this._options.icon.cloneNode(true) : undefined;
    }
    setIcon(icon) {
        this._options.icon = icon;
        this._onUpdatedEmitter.fire(this);
    }
    get toggledIcon() {
        return this._options.toggledIcon ? this._options.toggledIcon.cloneNode(true) : undefined;
    }
    get iconClass() {
        return this._options.iconName ? 'codicon codicon-' + this._options.iconName : '';
    }
    handle(event, target) {
        if (this._enabled)
            this._options.handle(this, event, target);
    }
    toggled() {
        return this._toggled;
    }
    setToggled(toggled) {
        this._toggled = toggled;
        this._onUpdatedEmitter.fire(this);
    }
    enabled() {
        return this._enabled;
    }
    setEnabled(enabled) {
        this._enabled = enabled;
        this._onUpdatedEmitter.fire(this);
    }
}
exports.Action = Action;
class ActionRegistry {
    constructor() {
        this._actions = new multimap_1.MultiMap();
    }
    createAction(options) {
        return new Action(options);
    }
    registerAction(targetTag, options) {
        const action = this.createAction(options);
        this._actions.set(targetTag, action);
        return action;
    }
    actionsForTags(actionTags) {
        const result = new Set();
        for (const tag of actionTags) {
            const actions = this._actions.get(tag);
            for (const a of actions)
                result.add(a);
        }
        return result;
    }
}
exports.actionRegistry = new ActionRegistry();
