// playwright-cli/lib/traceViewer/web/components/pwCheckbox.js
"use strict";
/*
  Copyright (c) Microsoft Corporation.
 
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at
 
      http://www.apache.org/licenses/LICENSE-2.0
 
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.PwCheckboxElement = void 0;
const events_1 = require("./events");
const dom_1 = require("./dom");
// @ts-ignore
const codicon_css_1 = require("!css-loader!../third_party/vscode/codicon.css");
// @ts-ignore
const pwCheckbox_css_1 = require("!css-loader!./pwCheckbox.css");
class PwCheckboxElement extends HTMLElement {
    constructor() {
        super();
        this._onCheckedEmitter = new events_1.EventEmitter();
        this.onChecked = this._onCheckedEmitter.event;
        this._shadowRoot = this.attachShadow({ mode: 'open', delegatesFocus: false });
        this._shadowRoot.appendChild(dom_1.dom `
      ${PwCheckboxElement.stylesFragment()}
      <div class="codicon" tabIndex=0></div>
      <span><slot></slot></span>
    `);
        this._checkboxElement = this._shadowRoot.querySelector('.codicon');
        this._checkboxElement.addEventListener('click', () => this.setChecked(!this.checked()));
        this._checkboxElement.classList.toggle('codicon-check', this.checked());
    }
    static get observedAttributes() {
        return ['checked'];
    }
    attributeChangedCallback(attr, oldValue, newValue) {
        if (attr === 'checked') {
            this._onCheckedEmitter.fire(this.checked());
            this._checkboxElement.classList.toggle('codicon-check', this.checked());
        }
    }
    checked() {
        return this.hasAttribute('checked');
    }
    setChecked(checked) {
        if (this.checked() === checked)
            return;
        if (checked)
            this.setAttribute('checked', '');
        else
            this.removeAttribute('checked');
    }
}
exports.PwCheckboxElement = PwCheckboxElement;
PwCheckboxElement.tagName = 'pw-checkbox';
PwCheckboxElement.styles = [codicon_css_1.default, pwCheckbox_css_1.default];
