// playwright-cli/lib/traceViewer/web/components/toolbarView.js
"use strict";
/*
  Copyright (c) Microsoft Corporation.
 
  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at
 
      http://www.apache.org/licenses/LICENSE-2.0
 
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
*/
Object.defineProperty(exports, "__esModule", { value: true });
exports.ToolbarView = void 0;
const dom_1 = require("./dom");
const actionRegistry_1 = require("./actionRegistry");
const listView_1 = require("./listView");
const events_1 = require("./events");
class ToolbarView {
    constructor() {
        this._enabled = true;
        this._disposables = [];
        this._listView = new listView_1.ListView(this, {
            orientation: 'horizontal'
        });
        this.element = dom_1.dom `<toolbar-view>${this._listView.element}</toolbar-view>`;
    }
    render(action) {
        const icon = action.toggled() ? action.toggledIcon || action.icon : action.icon;
        const toggledStyle = action.toggled() && (action.label || !action.toggledIcon) ? 'toggled' : '';
        const label = action.iconClass || icon ? '' : action.label;
        const item = dom_1.dom `
      <toolbar-item title="${action.title}" class="${this._enabled && action.enabled() ? '' : 'disabled'} ${toggledStyle}">
        <toolbar-icon class="${action.iconClass}">${icon || ''}</toolbar-icon>
        <toolbar-label>${label}</toolbar-label>
      </toolbar-item>
    `;
        item.addEventListener('mousedown', event => {
            event.preventDefault();
        });
        item.addEventListener('click', event => {
            event.stopPropagation();
            event.preventDefault();
            if (this._enabled)
                action.handle(event);
        });
        return item;
    }
    setEnabled(enabled) {
        this._enabled = enabled;
        this._listView.updateAll();
    }
    addItem(param) {
        const action = param instanceof actionRegistry_1.Action ? param : actionRegistry_1.actionRegistry.createAction(param);
        this._disposables.push(action.onUpdated(() => this._listView.update(action)));
        this._listView.append(action);
        return action;
    }
    clear() {
        this._listView.clear();
        events_1.Disposable.disposeAll(this._disposables);
    }
}
exports.ToolbarView = ToolbarView;
